src/go/rdctl/pkg/reg/reg.go

```diff
@@ -29,37 +29,60 @@ func escape(s string) string {
 //
 // Params:
 // pathParts: represents the registry path to the current item
-// v: the reflected value of the current field
+// structType: a reflection on the ServerSettingsForJSON struct, so we can determine how to interpret each field in the instance
+// value: the reflected value of the current field, based on a simple map[string]interface{} JSON-parse
 // jsonTag: the name of the field, used in json (and the registry)
+// path: a dotted representation of the fully-qualified name of the field
 //
 // Returns two values:
 //
 //	an array of lines representing the current value
 //	an error: the only non-nil error this function can return is when it encounters an unhandled value type
-func convertToRegFormat(pathParts []string, v reflect.Value, jsonTag string) ([]string, error) {
-	kind := v.Kind()
+func convertToRegFormat(pathParts []string, structType reflect.Type, value reflect.Value, jsonTag, path string) ([]string, error) {
+	kind := structType.Kind()
+	if value.Kind() == reflect.Interface && kind != reflect.Interface {
+		if value.IsNil() {
+			return nil, nil
+		}
+		return convertToRegFormat(pathParts, structType, value.Elem(), jsonTag, path)
+	}
+	if value.Kind() == reflect.Ptr {
+		return nil, fmt.Errorf("reg-file generation: got an unexpected pointer for %s value %v, expecting type %v", path, value, structType)
+	}
 	switch kind {
 	case reflect.Struct:
-		numFields := v.NumField()
-		scalarReturnedLines := make([]string, 0, numFields)
+		// This is very different from plist-processing of structs.
+		// In the plist world we want to preserve order of fields in the original JSON,
+		// because that's what the plutil converter does.
+		// In the registry world better to order the fields alphabetically
+		//
+		if value.Kind() != reflect.Map {
+			return nil, fmt.Errorf("expecting actual kind for a typed struct to be a map, got %v", value.Kind())
+		}
+		numTypedFields := structType.NumField()
+		sortedStructFields := utils.SortStructFields(structType)
+		scalarReturnedLines := make([]string, 0, numTypedFields)
 		nestedReturnedLines := make([]string, 0)
-		for i := 0; i < numFields; i++ {
-			fieldTag := v.Type().Field(i).Tag.Get("json")
-			fieldName := strings.Replace(fieldTag, ",omitempty", "", 1)
-			newRetLines, err := convertToRegFormat(append(pathParts, fieldName), v.Field(i), fieldName)
-			if err != nil {
-				return nil, err
-			}
-			if len(newRetLines) == 0 {
-				continue
-			}
-			// With the current schema, only structs may contain nested structs.
-			// If the first character of the first line is a '[' it's a struct. Otherwise it's a scalar.
-			// ']' placed here to appease my IDE's linter.
-			if newRetLines[0][0] == '[' {
-				nestedReturnedLines = append(nestedReturnedLines, newRetLines...)
-			} else {
-				scalarReturnedLines = append(scalarReturnedLines, newRetLines...)
+		for i := 0; i < numTypedFields; i++ {
+			compoundStructField := sortedStructFields[i]
```

Comment:
This would read better as:
```suggestion
		for _, compoundStructField := range sortedStructFields ```

Reply:
one of those cases where the index was needed before and got dropped.